### ISSUE TITLE
[P0] Dapr downgrade from 1.12.0 to 1.11.2 doesn't work

### ISSUE THREAD

**Issue by ASHIQUEMD** [dapr/dapr#7055]
## In what area(s)?

> /area runtime

## What version of Dapr?

> 1.12.0

## Expected Behavior

Downgrade to 1.11.2 should work seamlessly.

## Actual Behavior
When Dapr is downgraded from 1.12.0 to 1.11.2, dapr sidecar injector fails to inject the dapr sidecar due to tls issue.

```bash
# k logs dapr-sidecar-injector-59b4f68f6d-p6l28 -n dapr-system                                                     
time="2023-10-18T13:25:55.20080834Z" level=info msg="log level set to: info" instance=dapr-sidecar-injector-59b4f68f6d-p6l28 scope=dapr.injector type=log ver=1.11.2
time="2023-10-18T13:25:55.201074423Z" level=info msg="metrics server started on :9090/" instance=dapr-sidecar-injector-59b4f68f6d-p6l28 scope=dapr.metrics type=log ver=1.11.2
time="2023-10-18T13:25:55.201202215Z" level=info msg="starting Dapr Sidecar Injector -- version 1.11.2 -- commit 83ca1abb11ffe34211db55dcd36d96b94252827a" instance=dapr-sidecar-injector-59b4f68f6d-p6l28 scope=dapr.injector type=log ver=1.11.2
time="2023-10-18T13:25:55.207855631Z" level=warning msg="Unable to get SA tekton-pipelines:tekton-pipelines-controller UID" instance=dapr-sidecar-injector-59b4f68f6d-p6l28 scope=dapr.injector type=log ver=1.11.2
time="2023-10-18T13:25:55.207948173Z" level=info msg="Healthz server is listening on :8080" instance=dapr-sidecar-injector-59b4f68f6d-p6l28 scope=dapr.injector type=log ver=1.11.2
time="2023-10-18T13:25:55.208014465Z" level=info msg="Sidecar injector is listening on :4000, patching Dapr-enabled pods" instance=dapr-sidecar-injector-59b4f68f6d-p6l28 scope=dapr.injector type=log ver=1.11.2
2023/10/18 13:26:28 http: TLS handshake error from 10.244.0.1:58110: remote error: tls: bad certificate
```

** Note: Upgrading Dapr to 1.12.0 and app pods rollout restart, starts working again **

## Steps to Reproduce the Problem
```bash
# install latest dapr version
dapr init -k --runtime-version 1.12.0
# install your apps
k apply -f node.yaml

# downgrade dapr to 1.11.2
dapr upgrade -k --runtime-version 1.11.2

# rollout restart your applications
k rollout restart deployment/nodeapp
```
## Release Note

RELEASE NOTE:

**Comment by JoshVanL** [comment]
The reason why this occurs is because of [this](https://github.com/dapr/dapr/blob/18f2e3f98cf15db9b4d1051ed433a92fe553332a/charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_webhook_config.yaml#L46) if statement in the 1.11 helm chart. In 1.12 the injector mutating webhook CA bundle is the same as the rest of the Dapr cluster with the injector serving its identity certificate on the mutating endpoint. As such, the CA bundle is no longer being set through Helm (where the webhook CA bundle was also generated), and instead is [being set by the injector itself](https://github.com/dapr/dapr/blob/896e1b98425fbd6ef25d47caf7b6205100787c96/cmd/injector/main.go#L176).

The if statement in the 1.11 helm chart means that when downgrading from 1.12 to 1.11, the mutating webhook CA bundle will stay as the Dapr cluster CA rather than to the dedicated webhook CA. This causes a trust chain mismatch between the injector serving certificate and the trust anchor that the Kubernetes API server has been configured in the mutating webhook `caBundle` to trust.

To resolve this issue, I propose doing two things:

- Update the helm chart 1.11 with a patch release to remove this if statement entirely so the `caBundle` is templated as:  `    caBundle: {{ b64enc $ca.Cert }}`. I have confirmed that downgrading from 1.12 to 1.11 with this change works. It is also not clear to me why we want this if statement at all.
- Update the Dapr CLI to special case 1.12 to 1.11 downgrades to delete the mutating webhook just before applying the Helm chart. This will result in the if statement resolving to `false` and the `caBundle` will be set to the dedicated webhook CA. We should ensure we keep a copy of the mutating webhook before deletion, so that it can be re-created in the event of a failed upgrade, though we need to be sure that the Dapr CLI is correctly rolling back downgrades as the helm client is configured with `CleanupOnFail` rather than `Atomic`, see: https://github.com/helm/helm/blob/c5d78b89d1fe5adb3029f4407dab75dd40513631/pkg/action/upgrade.go#L90

Implementing both of these options ensures we fix the downgrade procedure for most versions and users.

**Comment by rabollin** [comment]
Downgrading to previous version is most commonly used scenario, and I would vote for option1.

**Comment by JoshVanL** [comment]
@rabollin I am not suggesting either or, I am suggesting both.

**Comment by mukundansundar** [comment]
@JoshVanL The main point for the second option as we discussed offline is a way to rollback any failures that we see on upgrade. I.e User updates from 1.11 to 1.12 and then finds some issue and wants to downgrade to 1.11. Do you think storing the 1.11 mutating web hook config across CLI calls is possible in that situation and will that be necessary?

Also do we need the CLI change if the helm chart change is done? Any particular reason for that?

Also that particular [line](https://github.com/dapr/dapr/blob/18f2e3f98cf15db9b4d1051ed433a92fe553332a/charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_webhook_config.yaml#L46) seems to have been added way back in 1.0 https://github.com/dapr/dapr/pull/2742

@yaron2 @artursouza Do you remember why this was added?

**Comment by JoshVanL** [comment]
> @JoshVanL The main point for the second option as we discussed offline is a way to rollback any failures that we see on upgrade. I.e User updates from 1.11 to 1.12 and then finds some issue and wants to downgrade to 1.11. Do you think storing the 1.11 mutating web hook config across CLI calls is possible in that situation and will that be necessary?

I don't see a reason that we would need to persist the config across calls, as the config will be installed on the second.

> Also do we need the CLI change if the helm chart change is done? Any particular reason for that?

It was to allow users to rollback to any `1.11` release, not just `1.11.5`, to ensure the most number of users don't run into this issue.

**Comment by ASHIQUEMD** [comment]
Hey @JoshVanL,

I validated the [patch](https://github.com/dapr/dapr/pull/7057) by upgrading and downgrading the Dapr version.
 
1.12.0 --> 1.11-patch - **worked fine**
1.11-patch - 1.12.0 - **worked fine**
1.12.0 --> 1.11-patch-->1.10.0 - **worked fine**
1.10.0->1.11-patch -> **worked fine**

1.12.0 --> 1.11.2 --> 1.10.0 -> **error**
1.12.0 --> 1.11.2 -> **error**
1.12.0 --> 1.10.0 -> **error**

**Comment by shivamkm07** [comment]
> Hey @JoshVanL,
> 
> I validated the [patch](https://github.com/dapr/dapr/pull/7057) by upgrading and downgrading the Dapr version.
> 
> 1.12.0 --> 1.11-patch - **worked fine** 1.11-patch - 1.12.0 - **worked fine** 1.12.0 --> 1.11-patch-->1.10.0 - **worked fine** 1.10.0->1.11-patch -> **worked fine**
> 
> 1.12.0 --> 1.11.2 --> 1.10.0 -> **error** 1.12.0 --> 1.11.2 -> **error** 1.12.0 --> 1.10.0 -> **error**

The CLI PR https://github.com/dapr/cli/pull/1359 fixes 1.12.0->1.11.X downgrade, but 1.12.0->1.10.X will still error if done directly. Through 1.11(1.12.0->1.11.X->1.10.X) it would work. Do we need to handle that scenario as well?

**Comment by JoshVanL** [comment]
@ASHIQUEMD @shivamkm07 We can add the patch to `1.10` as well, but downgrade across 2 minor Dapr versions is not supported, and considered best effort.

https://docs.dapr.io/operations/support/support-release-policy/#upgrade-paths
